Fix a fatal error not-terminating issue in Windows

hjyamauchi opened the pull request:
Instead of llvm.trap because it may not terminate the process and may leave the process state in limbo. For example, llvm.trap is turned into the ud2 instruction on X86 which can cause an infinite sigill on Windows.

hjyamauchi commented:
@compnerd

compnerd reviewed (changes requested):
This doesn't seem equivalent to me.  The `Builtin.int_trap()` lowers to `llvm.trap`, which based on the target may emit an instruction or a function call.  ~However, this now will lower to a function call to a function which may be entirely elided with `-DNDEBUG`~ (edit: nvm, I was mixing up `assert` and `abort`).

CC: @rjmccall @aschwaighofer

hjyamauchi committed:
Fix a fatal error not-terminating issue in Windows.

Fix the issue that fatal errors in certain cases don't terminate the
process and the process keeps running in Windows by disabling the
exception swallowing that supressed the illegal instruction exceptions
coming from llvm.trap.

compnerd commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -471,3 +471,32 @@ void swift::swift_abortDisabledUnicodeSupport() {
+// process may not terminate as expected on crashes. For example,
+// illegal instructions used by llvm.trap. Disable the exception
+// swallowing so that the error handling works as expected.
+__attribute__((constructor)) static void disableWinExceptionSwallowing() {
```
Please use the reserved spelling:

```suggestion
__attribute__((__constructor__))
static void disableWinExceptionSwallowing() {
```

I wonder if we should make this portable and use the same technique that we use in the registrar to ensure execution at the appropriate time.

I wonder if we can rename this something like `ConfigureExceptionPolicy`?

hjyamauchi replied:
Can you expand on "I wonder if we should make this portable and use the same technique that we use in the registrar to ensure execution at the appropriate time." ? What's the registrar?

compnerd replied:
https://github.com/apple/swift/blob/main/stdlib/public/runtime/SwiftRT-COFF.cpp#L100-L102

The use of constructor sections is how Windows does the constructor execution.

The registrar is the `swiftrt.obj` which registers the image with the runtime.

compnerd commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -471,3 +471,32 @@ void swift::swift_abortDisabledUnicodeSupport() {
+      }
+    }
+    BOOL Suppress = FALSE;
+    SetUserObjectInformationA(GetCurrentProcess(),
```
Do we want to suppress the exception even if we were unable to set the policy flag?

hjyamauchi replied:
It turned out that the info about SetProcessUserModeExceptionPolicy was out of date. It doesn't exist anymore. Removed it.

compnerd commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -471,3 +471,32 @@ void swift::swift_abortDisabledUnicodeSupport() {
+// On Windows, exceptions may be swallowed in some cases and the
+// process may not terminate as expected on crashes. For example,
+// illegal instructions used by llvm.trap. Disable the exception
+// swallowing so that the error handling works as expected.
```
Perhaps we should also document why a last chance vectored exception handler would be insufficient for this.

hjyamauchi replied:
Haven't tried it yet. Will try. But this works as is.

compnerd commented:
@swift-ci please test

compnerd commented:
Merging this for now, but I think that we should consider the portable alternative as well.
